### PR TITLE
Call out to firehose client when user toggles show code/block

### PR DESCRIPTION
**Pull request by pcardune**
Currently we are just logging the project_id and the state being toggled to. cc @poorvasingal for whether we want to log anything else?

### CODE REVIEW

**Comment by pcardune** [comment]
Also, I believe @ashercodeorg was going to add some code for logging user ids?

**Comment by ashercodeorg** [comment]
The system already logs the user_id, if provided. The documentation
(comment in firehouse.rb) is missing this, I'll send a PR to you to correct
the comment.

On Apr 24, 2017 2:37 PM, "Paul Carduner" <notifications@github.com> wrote:

Also, I believe @ashercodeorg <https://github.com/ashercodeorg> was going
to add some code for logging user ids?

**Comment by pcardune** [comment]
It will show up in redshift apparently.

**Comment by poorvasingal** [comment]
Sorry for the late reply.  Here are the set of things we'd want: 
- Log whenever student hits the toggle, along with what project or level they were on
- @ashercodeorg had also suggested logging what mode the student is in on page load of an App Lab / Game Lab level.  This will allow us generate useful graphs and find trends more easily.

File: dashboard/app/controllers/api/v1/users_controller.rb [hunk 1]
```diff
@@ -18,6 +20,16 @@ def get_using_text_mode
   def post_using_text_mode
     @user.using_text_mode = !!params[:using_text_mode].try(:to_bool)
     @user.save
+
+    FirehoseClient.instance.put_record(
+      'analysis-events',
+      {
+        study: 'project_block_and_text_switching',
+        event: @user.using_text_mode ? 'block_to_text' : 'text_to_block',
+        project_id: params[:project_id],
```
Comment (ashercodeorg) [inline dashboard/app/controllers/api/v1/users_controller.rb:30]: You can add `user_id: @user.id` here.

**Comment by pcardune** [comment]
So I added level_id and user_id, but now I'm getting the following error in development.log:

```
Started POST "/api/v1/users/me/using_text_mode" for 127.0.0.1 at 2017-04-26 17:03:17 +0000
Processing by Api::V1::UsersController#post_using_text_mode as URL_ENCODED_FORM
  Parameters: {"project_id"=>"qsGcjzrBN5myacU2kr69KA", "level_id"=>"custom", "using_text_mode"=>"true", "user_id"=>"me"}
Can't verify CSRF token authenticity.
Completed 422 Unprocessable Entity in 6ms (ActiveRecord: 0.0ms)

ActionController::InvalidAuthenticityToken - ActionController::InvalidAuthenticityToken:
```

This wasn't happening before and I don't see how adding additional params would cause this to occur. Did we recently change something about CSRF handling?

**Comment by ashercodeorg** [comment]
My guess is that `level_id` is required to be an integer. If that isn't it, I have no idea.

**Comment by davidsbailey** [comment]
I saw a similar error recently too before taking a different approach. The internet says that the following phrases from our source code pertain to the error you're seeing, so you could see if any of those lines have changed:
```
  skip_before_action :verify_authenticity_token
  protect_from_forgery
```
Incidentally it seems like your code would be more RESTful if you were doing a `put`, so you could try that instead -- there may be different security rules around `post` and `put`.

 @aoby or @wjordan , do you recognize the `ActionController::InvalidAuthenticityToken` error and/or know what to do about it?

**Comment by pcardune** [comment]
@davidsbailey Adding `skip_before_action :verify_authenticity_token` does seem to work.

**Pull request merged**